core/drivers/bcm_hwrng.c

```diff
@@ -44,29 +48,74 @@ static void bcm_hwrng_reset(void)
 			RNG_CTRL_MASK, RNG_CTRL_ENABLE);
 }
 
-uint32_t bcm_hwrng_read_rng(uint32_t *p_out, uint32_t words_to_read)
+static uint32_t do_rng_read(void *buf, ssize_t blen)
 {
-	uint32_t available_words = 0;
-	uint32_t num_words = 0;
-	uint32_t i = 0;
-	uint64_t timeout = timeout_init_us(RNG_TIMEOUT_US);
+	uint32_t available = 0;
+	ssize_t copied = 0;
+	ssize_t copy_len = 0;
+	union {
+		uint32_t word;
+		uint8_t  bytes[4];
+	} data;
+	uint64_t timeout;
 
 	assert(bcm_hwrng_base);
 
-	do {
-		available_words = io_read32(bcm_hwrng_base +
-					    RNG_FIFO_COUNT_OFFSET);
-		available_words = available_words & RNG_FIFO_COUNT_MASK;
-	} while (!available_words && !timeout_elapsed(timeout));
-
-	if ((available_words > 0) && (words_to_read > 0)) {
-		num_words =  MIN(available_words, words_to_read);
-		for (i = 0; i < num_words; i++)
-			p_out[i] = io_read32(bcm_hwrng_base +
-					     RNG_FIFO_DATA_OFFSET);
+	while (copied < blen) {
+		if (!available) {
+			timeout = timeout_init_us(RNG_TIMEOUT_US);
+			do {
+				available = io_read32(bcm_hwrng_base +
+						RNG_FIFO_COUNT_OFFSET);
+				available = available & RNG_FIFO_COUNT_MASK;
+				if (timeout_elapsed(timeout)) {
```

Comment:
`crypto_rng_read()` isn't supposed to timeout. If this is an unrecoverable error I guess we have no choice, but if it would help waiting a bit longer I believe we should do so instead.

Reply:
That would be a question for @bgooty, this patch is just refactoring original timeout handling to ensure correct FIFO underrun handling. I've not seen underrun during tests.

Reply:
Callers of `crypto_rng_read()` will error out if the function fails. So for instance if you read from RPMB you'll suddenly get an error. If it happens during boot in `rwp_init()` it will cause a panic. So I think this function should try very hard to only return errors on unrecoverable errors.

If it might help to wait a bit longer we should do so here, because there is not other place in the code better suited to handle this error. On the other hand, if it's quite likely that it is a permanent error then we should error out because there's a chance that there's still something useful left that the system can do.